src/Illuminate/Database/Eloquent/Casts/AsCollection.php

```diff
@@ -19,13 +19,28 @@ public static function castUsing(array $arguments)
         return new class implements CastsAttributes {
             public function get($model, $key, $value, $attributes)
             {
-                return isset($attributes[$key]) ? new Collection(json_decode($attributes[$key], true)) : null;
+                $items = isset($attributes[$key]) ? json_decode($attributes[$key], true) : [];
+
+                return new Collection($items);
             }
 
             public function set($model, $key, $value, $attributes)
             {
+                if (is_null($model->getRawOriginal($key)) && $this->isEmpty($value)) {
+                    return [$key => null];
+                }
+
                 return [$key => json_encode($value)];
             }
+
+            protected function isEmpty($value)
+            {
+                if ($value instanceof Collection) {
+                    return $value->isEmpty();
+                }
+
+                return $value === null;
+            }
```

Comment:
Whats the point of checking the original? Would this work instead?
```suggestion
            public function set($model, $key, $value, $attributes)
            {
                if (is_null($value) || ($value instanceof Collection && $value->isEmpty())) {
                    return [$key => null];
                }

                return [$key => json_encode($value)];
            }
```